### PR TITLE
GAWB - 2627 - Allow multiple clusters with same name

### PR DESCRIPTION
**Pull request by vkumra-broad**
Updated the unique constraint to include destroyed date. Added default destroyed date on cluster create. Only one cluster is active at a time. 

Have you read [CONTRIBUTING.md](../CONTRIBUTING.md) lately? If not, do that first.

I, the developer opening this PR, do solemnly pinky swear that:

- [ ] I've documented my API changes in Swagger

In all cases:

- [ ] Get two thumbsworth of review and PO signoff if necessary
- [ ] Verify all tests go green
- [ ] Squash and merge; you can delete your branch after this
- [ ] Test this change deployed correctly and works on dev environment after deployment

### CODE REVIEW

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/ClusterComponent.scala [hunk 1]
```diff
@@ -72,8 +74,14 @@ trait ClusterComponent extends LeoComponent {
       }
     }
 
-    def getByName(project: GoogleProject, name: ClusterName): DBIO[Option[Cluster]] = {
-      clusterQueryWithLabels.filter { _._1.googleProject === project.string }.filter { _._1.clusterName === name.string }.result map { recs =>
+    def getActiveClusterByName(project: GoogleProject, name: ClusterName): DBIO[Option[Cluster]] = {
+      clusterQueryWithLabels.filter { _._1.googleProject === project.string }.filter { _._1.clusterName === name.string }.filter{_._1.destroyedDate === Timestamp.from(Instant.parse(dummyDate))}.result map { recs =>
```
Comment (helgridly) [inline src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/ClusterComponent.scala]: Can you break this out onto multiple lines for readability? Something like:

```
clusterQueryWithLabels
.filter { _._1.googleProject === project.string }
.filter { _._1.clusterName === name.string }
.filter{_._1.destroyedDate === Timestamp.from(Instant.parse(dummyDate))}
.result map { recs =>
```

Comment (helgridly) [reply src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/ClusterComponent.scala]: (and/or you can just put all the conditionals in a single filter)

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/ClusterComponent.scala [hunk 2]
```diff
@@ -189,12 +193,18 @@ trait ClusterComponent extends LeoComponent {
...
         clusterRecord.hostIp map IP,
         clusterRecord.createdDate.toInstant,
-        clusterRecord.destroyedDate map { _.toInstant },
+        getDestroyedDate(clusterRecord.destroyedDate),
         labels,
         clusterRecord.jupyterExtensionUri flatMap { GcsPath.parse(_).toOption }
       )
     }
 
+    private def getDestroyedDate(destroyedDate:Timestamp): Option[Instant] = {
```
Comment (helgridly) [inline src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/ClusterComponent.scala]: I think this (and the sentinel magic-value destroyed date) should probably be functions on `object Cluster` instead -- but there might be disagreement here

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala [hunk 3]
```diff
@@ -61,14 +61,20 @@ class LeonardoService(protected val dataprocConfig: DataprocConfig, gdDAO: Datap
     }
   }
 
-  def getClusterDetails(googleProject: GoogleProject, clusterName: ClusterName): Future[Cluster] = {
+  /*def getClusterDetails(googleProject: GoogleProject, clusterName: ClusterName): Future[Cluster] = {
     dbRef.inTransaction { dataAccess =>
       getCluster(googleProject, clusterName, dataAccess)
     }
+  }*/
```
Comment (helgridly) [inline src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala]: remove

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/ClusterComponent.scala [hunk 4]
```diff
@@ -72,8 +74,14 @@ trait ClusterComponent extends LeoComponent {
...
 
-    def getByName(project: GoogleProject, name: ClusterName): DBIO[Option[Cluster]] = {
-      clusterQueryWithLabels.filter { _._1.googleProject === project.string }.filter { _._1.clusterName === name.string }.result map { recs =>
+    def getActiveClusterByName(project: GoogleProject, name: ClusterName): DBIO[Option[Cluster]] = {
+      clusterQueryWithLabels.filter { _._1.googleProject === project.string }.filter { _._1.clusterName === name.string }.filter{_._1.destroyedDate === Timestamp.from(Instant.parse(dummyDate))}.result map { recs =>
+        unmarshalClustersWithLabels(recs).headOption
+      }
+    }
+
+    def getDeletingClusterByName(project: GoogleProject, name: ClusterName): DBIO[Option[Cluster]] = {
```
Comment (helgridly) [inline src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/ClusterComponent.scala]: just checking: this is only used by tests?

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/ClusterComponent.scala [hunk 5]
```diff
@@ -40,7 +41,7 @@ trait ClusterComponent extends LeoComponent {
     def status =                column[String]            ("status",                O.Length(254))
     def hostIp =                column[Option[String]]    ("hostIp",                O.Length(254))
     def createdDate =           column[Timestamp]         ("createdDate",           O.SqlType("TIMESTAMP(6)"))
-    def destroyedDate =         column[Option[Timestamp]] ("destroyedDate",         O.SqlType("TIMESTAMP(6)"))
+    def destroyedDate =         column[Timestamp]         ("destroyedDate",         O.SqlType("TIMESTAMP(6)"))
```
Comment (rtitle) [inline src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/ClusterComponent.scala]: So is `destroyedDate` becoming not-null? Is there a migration for this?

Also what to do about _existing_ clusters in the DB who have a null `destroyedDate`?

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/ClusterComponent.scala [hunk 6]
```diff
@@ -50,6 +51,7 @@ trait ClusterComponent extends LeoComponent {
 
   object clusterQuery extends TableQuery(new ClusterTable(_)) {
 
+    private lazy val dummyDate:String = "1970-01-01T00:00:01.000Z"
```
Comment (rtitle) [inline src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/ClusterComponent.scala]: It looks like this is always parsed into an `Instant` -- could the constant just be `final val dummyDate =Instant.ofEpochMilli(1)`?

I don't think you need the `lazy` either.

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/ClusterComponent.scala (hunk 1 above)
Comment (rtitle) [reply src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/ClusterComponent.scala]: +1 for shorter lines

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/ClusterComponent.scala [hunk 7]
```diff
@@ -87,15 +95,11 @@ trait ClusterComponent extends LeoComponent {
     def markPendingDeletion(googleId: UUID): DBIO[Int] = {
       clusterQuery.filter(_.googleId === googleId)
         .map(c => (c.destroyedDate, c.status, c.hostIp))
-        .update((Option(Timestamp.from(java.time.Instant.now())), ClusterStatus.Deleting.toString, None))
+        .update(Timestamp.from(java.time.Instant.now()), ClusterStatus.Deleting.toString, None)
```
Comment (rtitle) [inline src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/ClusterComponent.scala]: Minor, `java.time.Instant` is imported so just `Instant.now()`

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/ClusterComponent.scala (hunk 2 above)
Comment (rtitle) [reply src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/ClusterComponent.scala]: 🤷‍♂️ I like it here -- in `Cluster` we abstract away the fact that we have a magic sentinel value in the database. The sentinel-handling code is private to `ClusterComponent`.

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/ClusterComponent.scala [hunk 8]
```diff
@@ -189,12 +193,18 @@ trait ClusterComponent extends LeoComponent {
...
         clusterRecord.createdDate.toInstant,
-        clusterRecord.destroyedDate map { _.toInstant },
+        getDestroyedDate(clusterRecord.destroyedDate),
         labels,
         clusterRecord.jupyterExtensionUri flatMap { GcsPath.parse(_).toOption }
       )
     }
 
+    private def getDestroyedDate(destroyedDate:Timestamp): Option[Instant] = {
+      if(destroyedDate.toInstant != Instant.parse((dummyDate)))
```
Comment (rtitle) [inline src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/ClusterComponent.scala]: You have an extra set of parens :)

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/ClusterComponent.scala (hunk 4 above)
Comment (vkumra-broad) [reply src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/ClusterComponent.scala]: Yes, I changed the `getByName` to `getActiveClusterByName`, so had to create another one just for the tests to return the deleting cluster. 

Should `getActiveClusterByName` return a deleting cluster?

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/ClusterComponent.scala (hunk 5 above)
Comment (helgridly) [reply src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/ClusterComponent.scala]: I'm willing to sweep this under the rug and deal with it by dropping the dev db for now -- but I agree that this isn't sustainable and we should be migrating in our changesets too.

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/ClusterComponent.scala (hunk 4 above)
Comment (helgridly) [reply src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/ClusterComponent.scala]: that would be incredibly confusing. i vote no :) it wasn't a criticism, just confirming

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/ClusterComponent.scala (hunk 5 above)
Comment (vkumra-broad) [reply src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/ClusterComponent.scala]: Added the changeset for this.

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/ClusterComponent.scala (hunk 2 above)
Comment (vkumra-broad) [reply src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/ClusterComponent.scala]: I prefer this here as well, for the same reason as above.

**Pull request merged**